Libraries/LibTLS/TLSv12.cpp

```diff
@@ -0,0 +1,516 @@
+
+    return res;
+}
+
+void TLSv12::consume(const ByteBuffer& record)
+{
+    if (m_context.critical_error) {
+        dbg() << "There has been a critical error (" << (i8)m_context.critical_error << "), refusing to continue";
+        return;
+    }
+
+    if (record.size() == 0) {
+        return;
+    }
+
+#ifdef TLS_DEBUG
+    dbg() << "Consuming " << record.size() << " bytes";
+#endif
+
+    m_context.message_buffer.append(record.data(), record.size());
+
+    size_t index { 0 };
+    size_t buffer_length = m_context.message_buffer.size();
+
+    size_t size_offset { 3 }; // read the common record header
+    size_t header_size { 5 };
+#ifdef TLS_DEBUG
+    dbg() << "message buffer length " << buffer_length;
+#endif
+    while (buffer_length >= 5) {
+        auto length = convert_between_host_and_network(*(u16*)m_context.message_buffer.offset_pointer(index + size_offset)) + header_size;
+        if (length > buffer_length) {
+            dbg() << "Need more data: " << length << " | " << buffer_length;
+            break;
+        }
+        auto consumed = handle_message(m_context.message_buffer.slice_view(index, length));
+        if (consumed > 0)
+            dbg() << "consumed " << (size_t)consumed << " bytes";
+        else
+            dbg() << "error: " << (int)consumed;
+
+        if (consumed != (i8)Error::NeedMoreData) {
+            if (consumed < 0) {
+                dbg() << "Consumed an error: " << (int)consumed;
+                if (!m_context.critical_error)
+                    m_context.critical_error = (i8)consumed;
+                m_context.error_code = (Error)consumed;
+                break;
+            }
+        } else {
+            continue;
+        }
+
+        index += length;
+        buffer_length -= length;
+        if (m_context.critical_error) {
+            dbg() << "Broken connection";
+            m_context.error_code = Error::BrokenConnection;
+            break;
+        }
+    }
+    if (m_context.error_code != Error::NoError && m_context.error_code != Error::NeedMoreData) {
+        dbg() << "consume error: " << (i8)m_context.error_code;
+        m_context.message_buffer.clear();
+        return;
+    }
+
+    if (index) {
+        m_context.message_buffer = m_context.message_buffer.slice(index, m_context.message_buffer.size() - index);
+    }
+}
+
+void TLSv12::ensure_hmac(size_t digest_size, bool local)
+{
+    if (local && m_hmac_local)
+        return;
+
+    if (!local && m_hmac_remote)
+        return;
+
+    auto hash_kind = Crypto::Hash::HashKind::None;
+
+    switch (digest_size) {
+    case Crypto::Hash::SHA1::DigestSize:
+        hash_kind = Crypto::Hash::HashKind::SHA1;
+        break;
+    case Crypto::Hash::SHA256::DigestSize:
+        hash_kind = Crypto::Hash::HashKind::SHA256;
+        break;
+    case Crypto::Hash::SHA512::DigestSize:
+        hash_kind = Crypto::Hash::HashKind::SHA512;
+        break;
+    default:
+        break;
```

Comment:
If we didn't detect the digest size, should we do something ? Or is the passthrough with `hash_kind = None` enough ?

Reply:
I guess we'll log something at least